### ISSUE TITLE
onSubmit prop not working with React.cloneElement?

### ISSUE THREAD

**Issue by hunterc** [erikras/redux-form#294]
I have a form component, `SampleForm`, that I am passing an `onSubmit` prop to.
I am passing my form as a child to a wrapper `Card` component. The wrapper component clones the form and overrides the `onSubmit` prop with a composed version that closes the card.

Issue I'm running into is that for some reason the `onSubmit` prop isn't making it to the form. If I change the prop name to anything else it shows up.

Example:
```
<Card>
  <div>...</div>
  <SampleForm
    onSubmit={() => console.log('submit')}
  />
</Card>
```

Card.jsx bits of interest
```
let form = this.props.children[1];
let clonedForm = React.cloneElement(form, {
  onSubmit: (::this.composeSubmit(form.props))
});
```

```
composeSubmit({ onSubmit }) {
  return (data) => {
    ::this.closeCard();
    onSubmit(data);
  };
}
```

When I check the props of my form component the `onSubmit` prop is undefined.

**Comment by erikras** [comment]
What does this print out?
```js
let form = this.props.children[1];
console.log(form.props);
```
Is there an `onSubmit` in there?

**Comment by hunterc** [comment]
![screen shot 2015-11-18 at 10 00 42 am](https://cloud.githubusercontent.com/assets/2365571/11244366/475fcb2e-8ddb-11e5-949f-6f2dd5f3c472.png)
The chrome collapsed preview doesn't show onSubmit but when I expand it shows up.

**Comment by erikras** [comment]
I'm not understanding the problem. Can you rephrase it?

**Comment by hunterc** [comment]
Sure sorry for any confusion. The chrome bit was just an observation.

The issue I'm having is that while the props are showing up on the original form, this.props.children[1], when I clone the element with `React.cloneElement`, the `onSubmit` prop is missing.

**Comment by erikras** [comment]
I have never used `React.cloneElement`, so I am speaking from a position of ignorance, but maybe, just maybe, `React.cloneElement` is cloning the virtual dom (what React Components output), not the _React Components_ themselves, so you're losing all of the logic built into the `ReduxForm` component? Is that a possibility?

**Comment by hunterc** [comment]
Pretty sure it just operates on the component itself. [source](https://github.com/facebook/react/blob/0b29035484f428cb56e7e1c04a88f66ac020d1d4/src/isomorphic/classic/element/ReactElement.js#L215)

Thought onSubmit might be a reserved word for the clone but doesn't appear to be.

**Comment by hunterc** [comment]
@erikras  Could it be an issue in [here](https://github.com/erikras/redux-form/blob/c791ea910200636dbe819adfe7ced5f833da4609/src/createHigherOrderComponent.js#L97)?

I see you are grabbing `onSubmit` from the props but not passing it to the `WrappedComponent`.

**Comment by erikras** [comment]
Okay, your PR made it clear the behavior that you are expecting, but I'm unclear as to why you want that?

This:
```js
render() {
  const {handleSubmit} = this.props;
  return <form onSubmit={handleSubmit}></form>;
}
```
is the equivalent of
```js
render() {
  const {handleSubmit, onSubmit} = this.props;
  return <form onSubmit={handleSubmit(onSubmit)}></form>;
}
```
Are you wanting to do some pre-processing of the data before submitting it? Why do you need `onSubmit`?

**Comment by hunterc** [comment]
In the docs it mentions that you can specify an `onSubmit` prop on a form component either when rendering or through the redux `mapStateToProps` binding.

Since you are allowed to specify an `onSubmit` prop on a form component, it just seems to me that you should be able to pull it from the Wrapped Component.

**Comment by erikras** [comment]
My primary concern is that it might lead to some misinformed soul to do something like...
```js
render() {
  const {onSubmit} = this.props;
  return <form onSubmit={onSubmit}></form>;
}
```
...which looks perfectly fine, but wouldn't work at all.

What is it that you need to do _between_ `handleSubmit` and `onSubmit`?

**Comment by hunterc** [comment]
That's a good point. Maybe change the name of the `onSubmit` prop on the Wrapped Component?

My use case is that I am passing an instance of my form to another component as a child. On that instance I specified an `onSubmit` prop to hook into the form submission to do something  like update another reducer based on the form data.

The component I am passing my form to needs to hook into the form being submitted so it can perform some UI change, flipping a card over for example.

I normally do this by just composing the functions together and cloning the component, passing in the composed function where the old hook was, `onSubmit` in this case.

**Comment by erikras** [comment]
OH!! Now your original code makes some sense. Wow, that's an awkward design, but I can see how it would be nice for the form component to be ignorant of whether or not it's being housed in a card.

It might be worth taking a step back and seeing if the flow of data in Redux could help. e.g. Could somehow the reducer holding the "which cards are flipped?" state listen for some part of the action dispatched by `onSubmit`?

Even if we could rename `onSubmit`, it still wouldn't work, because `React.cloneElement` is (I think!) working on the virtual dom. If we called it `originalOnSubmit`, you couldn't do:
```js
 React.cloneElement(form, {
  onSubmit: (::this.composeSubmit(form.props)) // <---- would never actually get to ReduxForm component
});

composeSubmit({ originalOnSubmit }) {
  return (data) => {
    ::this.closeCard();
    originalOnSubmit(data);
  };
}
```

**Comment by hunterc** [comment]
@erikras any more thoughts on this? I guess my main point is it just feels awkward to define a prop on a component and then not directly have access to it from inside that component.

**Comment by erikras** [comment]
There is a callback hack to let you call a method on a component from one of its ancestors. Have a look at https://github.com/erikras/redux-form/issues/173#issuecomment-157844254. I can explain further if it's not immediately clear how to apply it to your problem.

**Comment by shanewho** [comment]
I just ran into this. My problem is I have a component called MyForm that contains a redux form. So I want to do something like this:

```
render (
   <some>
      <stuff>
         <MyForm onSubmit={this.blar}
```

This works fine, except I want to do some preprocessing inside of MyForm. Kind of like this:

```
  onSubmit(values, next) {
    values.type = this.state.type;
    this.props.onSubmit(values); //<-- I can't do this
  }

  render() {
    const {fields: {name, location, type}, handleSubmit} = this.props;
    return (
      <Form action='/rules' onSubmit={handleSubmit(this.onSubmit.bind(this))}>
```

The workaround is just to use something that isn't named onSubmit (onSubmit2, ha), but that is kind of odd since onSubmit is the name that makes the most sense.

**Comment by ooflorent** [comment]
As @shanewho said, the workaround is to use another property name than `onSubmit` prior v6.
Since v6, `onSubmit` is passed down.
